hoomd/mpcd/SphereGeometry.h

```diff
@@ -0,0 +1,195 @@
+         *       according to the appropriate bounce back rule, and the integration time \a dt is decreased to the
+         *       amount of time remaining.
+         */
+        HOSTDEVICE bool detectCollision(Scalar3& pos, Scalar3& vel, Scalar& dt) const
+            {
+
+            /*
+             * Detect if particle has left the box, and try to avoid branching or absolute value calls. The sign used
+             * in calculations is 1 if the particle is out-of-bounds in the radial direction, 0 otherwise.
+             *
+             * We intentionally use > / < rather than >= / <= to make sure that spurious collisions do not get detected
+             * when a particle is reset to the boundary location. A particle landing exactly on the boundary from the bulk
+             * can be immediately reflected on the next streaming step, and so the motion is essentially equivalent up to
+             * an epsilon of difference in the channel width.
+             */
+            double r2 = pos.x*pos.x + pos.y*pos.y + pos.z*pos.z;
+            const char sign = (r2 > R*R);
+            // exit immediately if no collision is found or particle is not moving normal (radial direction) to the wall
+            // (since no new collision could have occurred if there is no normal motion)
+            /*
+            r(t) = sqrt(x**2 + y**2 + z**2)
+            radial velocity, r_dot = (x*x_dot + y*y_dot + z*z_dot)/r(t)
+            */
+            float vr = (pos.x*vel.x + pos.y*vel.y + pos.z*vel.z)/sqrt(r2);
+            if (sign == 0 || vr == Scalar(0))
+               {
+               dt = Scalar(0);
+               return false;
+               }
+
+            // Find the point of contact when the particle is just leaving the surface
+            /*
+            * Calculating the time spent outside-bounds requires the knowledge of the point of contact on the boundary.
+            * The point of contact can be calculated using geometrical considerations.
+            *
+            * We know the following quantities,
+            *    1. r(t+del_t)      (the point outside the sphere)
+            *    2. v(t)            (previous velocity)
+            *
+            * Assuming 'r*' is the point of contact on the spherical shell and dx* is the distance travelled by the
+            * particle outside the boundary.
+            *
+            * r* = r(t+del_t) + dx* * v(t)/|v|          (the -ve signs cancel out in the 2nd term)
+            * and |r|**2 = R**2
+            *
+            * Solving the above equations,
+            * dx* = -r(t+del_t).v(t)/|v| +- sqrt((r(t+del_t).v(t))**2 - |r(t+del_t)|**2 + R**2)
+            * consequently, dt = dx* / |v|
+            */
+            double mod_v,r_dot_v_hat,dx_;
+
+            mod_v = sqrt(vel.x*vel.x + vel.y*vel.y + vel.z*vel.z);
+            r_dot_v_hat = (pos.x*vel.x + pos.y*vel.y + pos.z*vel.z)/mod_v;
```

Comment:
You can recycle the variable you already computed. Now is also the time to call the square root (exactly once), and you should use the `fast::` version from the HOOMDMath header.
```suggestion
            Scalar v2 = dot(vel,vel);
            r_dot_v_hat = vr/fast::sqrt(r2*v2);
```
Note that this raises one potential problem case: what happens if v2 is zero? I would recommend moving up the calculation of v2 and adding this case to the early-exit statement above.

Reply:
You still have to treat the case r2*v2=0, since you could otherwise have a division by zero in the calculation of r_dot_v_hat. Such a case could occur if a solvent particle is located at the origin (r2=0) or if its velocity is zero (v2=0). These cases are unlikely, but could still happen.